src/Microsoft.AspNetCore.Hosting/Internal/WebHost.cs

```diff
@@ -267,23 +272,51 @@ private void EnsureServer()
             }
         }
 
-        public void Dispose()
+        public async Task StopAsync(CancellationToken cancellationToken)
         {
+            if (_stopped)
+            {
+                return;
+            }
+            _stopped = true;
+
             _logger?.Shutdown();
 
+            if (!cancellationToken.CanBeCanceled)
+            {
+                cancellationToken = new CancellationTokenSource(Options.ShutdownTimeout).Token;
```

Comment:
We should only do this if there is a timeout right? How do you disable the timeout?